### ISSUE TITLE
Provide options to filter packages by release date for reproducing old envts?

### ISSUE THREAD

**Issue by jbednar** [conda/conda#6947]
**I'm submitting a...**
  - [x] feature request

`conda env create` normally gives the newest version of each package compatible with the environment specification, which is a good default.

However, environment files typically only loosely specify the package versions needed, because users rarely know which version of a package introduced the features they rely on, and because they cannot anticipate (semver notwithstanding) which packages will have API changes or new bugs in the future. Thus if I try to work with code from a colleague/student/collaborator/etc. that's more than a few weeks old, the chance that `conda env create` will install the same set of package versions as when they made their own environment is very low.  In practice this makes it very difficult to reproduce their work (or even my own old work),  unless I happen to have saved  the output of `conda env export`.

So that people can reproduce (at least roughly) what an environment created on date X in the past would have contained, can `conda` have an option to filter out all packages released after date X?  As conda becomes more integrated into peoples' workflows and more conda environments start being strewn about codebases, I think this issue will come up more and more frequently for users...

**Comment by goanpeca** [comment]
@jbednar how would this feature look, do you have a CLI example in mind?

```
conda install soma-package --date-something 2018-01-01
conda install soma-package[date<="2018-01-01"]
```

**Comment by jbednar** [comment]
I'd be happy with any API that would do the job, as long as it accepts a human-readable date (not milliseconds since the epoch :-).  But note that I'm suggesting this for ``conda env create``, i.e. for creating an environment from an underspecified file (as they always are), not for ``conda install`` of a single package (which can be specified easily already).

So the equivalent would be something like:

```
conda env create --no-later-than      2018-01-01
conda env create --package-date-limit 2018-01-01
conda env create --max-date           2018-01-01
conda env create --up-to              2018-01-01
conda env create --as-of              2018-01-01
```

I'm not expecting the date to include "<=", but that's just because I can't think of a use for saying that only packages *newer* than a certain date can be used (">="), or that only packages *from* a certain date can be used ("==").

**Comment by jbednar** [comment]
Yes, this issue still applies to all versions of conda on all platforms, unless I missed the introduction of a new feature somewhere along the way. Still would be _so_ useful!

**Comment by Jacob-Barhak** [comment]
This comment became stale which makes no sense. Once of the ideas with conda is to make work more reproducible - this is especially important for science which today is lacking in reproducibility and therefore credibility is damaged. 
I personally am supporting legacy code that I need to reinstall occasionally. Every few years there is a small hiccup that requires me to deal with packaging. 
These days when I install legacy code, conda looks for new libraries that did not exist in the time when my code version was created. Those libraries cause broken environments. 
The solutions @jbednar suggested should not be hard to implement - just exclude all library versions that do not match a date range before feeding those to the solver. 
The only thing I may add is perhaps a beginning date that should be a soft requirement. If I am recreating a package installation from 2015 I don't want packages from 2010 be used unless there is no other solution. Because if I installed it in 2015 and it worked, it would probably use recent versions to 2015. 
Here are examples:
`conda install package --min-date=2018-01-01 --max-date=2019-01-01`
or 
`conda install package --min-date=2018-01-01 --max-date=2019-01-01 --soft-date-restriction`

The latter one will install packages out of date only if possible.

I understand that using the most recent libraries up to a certain date that match the requirements is still not a guarantee of reproducibility. Yet I do want conda to try to reproduce what would have worked in a specific date in the past. If its not possible to cap the date, it is perhaps possible to put a penalty on the solver for selecting libraries farther from a specific date... 

something like:
`conda install package --best-date=2018-01-01`
 
Each possible solution suggested has pros and cons, yet without a solution that looks at a date, the promise of reproducibility in conda is a bad promise in many cases. People don't understand versions, people understand dates so if one wants a more human approach, please try to include a date constraint in conda. Even if it helps in some cases and not all cases, it is a blessing.

**Comment by jbednar** [comment]
uv calls this option ["--exclude-newer"](https://github.com/astral-sh/uv?tab=readme-ov-file#time-restricted-reproducible-resolutions), which is as good a name as any of my suggestions above.
